File: kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define EPSILON 0.001
#define DEFAULT_ITERATIONS 200
#define MAX_ITERATIONS 1000

double distance(double *point1, double *point2, int d);
int areBiggerThanEpsilon(double *arr, int size);
int is_integer(char *str);

int main(int argc, char **argv)
{
    int k, n, d, iter;
    double **vectors_arr;
    int i, j;
    int count = 0;
    double **centroids;
    int *closest_centroid_for_vector;
    int iteration = 0;
    double *delta_centroids;
    int point_index;
    double closest_centroid_distance;
    int centroid_index;
    double curr_d;
    double *sum;
    double *new_centroid;

    if (argc != 4 && argc != 5)
    {
        printf("Invalid number of arguments!\n");
        return 1;
    }

    if (!(is_integer(argv[1])))
    {
        printf("Invalid number of clusters!\n");
        return 1;
    }
    if (!(is_integer(argv[2])))
    {
        printf("Invalid number of points!\n");
        return 1;
    }
    if (!(is_integer(argv[3])))
    {
        printf("Invalid dimension of point!\n");
        return 1;
    }
    if (atoi(argv[1]) > 1 && atoi(argv[1]) < atoi(argv[2]))
    {
        k = atoi(argv[1]);
    }
    else
    {
        printf("Invalid number of clusters!\n");
        return 1;
    }
    if (atoi(argv[2]) > 1)
    {
        n = atoi(argv[2]);
    }
    else
    {
        printf("Invalid number of points!\n");
        return 1;
    }
    if (atoi(argv[3]) >= 1)
    {
        d = atoi(argv[3]);    
    }
    else
    {
        printf("Invalid dimension of point!\n");
        return 1;
    }
    if (argc == 4)
    {
        iter = DEFAULT_ITERATIONS;
    }
    else
    {
        if (!(is_integer(argv[4])))
        {
            printf("Invalid maximum iteration!\n");
            return 1;
        }
        if (atoi(argv[4]) > 1 && atoi(argv[4]) < MAX_ITERATIONS)
        {
            iter = atoi(argv[4]);
        }
        else
        {
            printf("Invalid maximum iteration!\n");
            return 1;
        }
    }

    /* Allocate memory for vectors */
    vectors_arr = (double **)malloc(n * sizeof(double *));
    for (i = 0; i < n; i++)
    {
        vectors_arr[i] = (double *)malloc(d * sizeof(double));
        for (j = 0; j < d; j++)
        {
            scanf("%lf,", &vectors_arr[i][j]);
        }
    }

    /* Allocate memory for centroids */
    centroids = (double **)malloc(k * sizeof(double *));
    for (i = 0; i < k; i++)
    {
        centroids[i] = (double *)malloc(d * sizeof(double));
        for (j = 0; j < d; j++)
        {
            centroids[i][j] = vectors_arr[i][j];
        }
    }

    closest_centroid_for_vector = (int *)malloc(n * sizeof(int));

    for (i = 0; i < n; i++)
    {
        double min_distance = distance(vectors_arr[i], centroids[0], d);
        closest_centroid_for_vector[i] = 0;
        for (j = 1; j < k; j++)
        {
            double curr_distance = distance(vectors_arr[i], centroids[j], d);
            if (curr_distance < min_distance)
            {
                closest_centroid_for_vector[i] = j;
                min_distance = curr_distance;
            }
        }
    }

    delta_centroids = (double *)malloc(k * sizeof(double));
    for (i = 0; i < k; i++)
    {
        delta_centroids[i] = 1.0;
    }

    sum = (double *)malloc(d * sizeof(double));
    for (i = 0; i < d; i++)
    {
        sum[i] = 0.0;
    }

    new_centroid = (double *)malloc(d * sizeof(double));
    for (i = 0; i < d; i++)
    {
        new_centroid[i] = 0.0;
    }

    while (areBiggerThanEpsilon(delta_centroids, k) && iteration < iter)
    {
        for (point_index = 0; point_index < n; point_index++)
        {
            closest_centroid_distance = distance(vectors_arr[point_index], centroids[closest_centroid_for_vector[point_index]], d);
            for (centroid_index = 0; centroid_index < k; centroid_index++)
            {
                curr_d = distance(vectors_arr[point_index], centroids[centroid_index], d);
                if (curr_d < closest_centroid_distance)
                {
                    closest_centroid_for_vector[point_index] = centroid_index;
                    closest_centroid_distance = curr_d;
                }
            }
        }
        for (centroid_index = 0; centroid_index < k; centroid_index++)
        {
            for (i = 0; i < d; i++)
            {
                sum[i] = 0.0;
            }
            for (i = 0; i < d; i++)
            {
                new_centroid[i] = 0.0;
            }
            count = 0;
            for (i = 0; i < n; i++)
            {
                if (closest_centroid_for_vector[i] == centroid_index)
                {
                    count++;
                    for (j = 0; j < d; j++)
                    {
                        sum[j] += vectors_arr[i][j];
                    }
                }
            }
            for (j = 0; j < d; j++)
            {
                new_centroid[j] = sum[j] / count;
            }
            delta_centroids[centroid_index] = distance(centroids[centroid_index], new_centroid, d);
            for (i = 0; i < d; i++)
            {
                centroids[centroid_index][i] = new_centroid[i];
            }
        }
        iteration++;
    }

    /* Print centroids */
    for (i = 0; i < k; i++)
    {
        for (j = 0; j < d; j++)
        {
            printf("%.4f", centroids[i][j]);
            if (j < d - 1)
            {
                printf(",");
            }
            else
            {
                printf("\n");
            }
        }
    }

    /* Free allocated memory */
    for (i = 0; i < k; i++)
    {
        free(centroids[i]);
    }
    free(centroids);
    free(closest_centroid_for_vector);
    free(delta_centroids);
    free(sum);
    free(new_centroid);
    for (i = 0; i < n; i++)
    {
        free(vectors_arr[i]);
    }
    free(vectors_arr);

    return 0;
}

double distance(double *point1, double *point2, int d)
{
    double sum = 0.0;
    int i;
    for (i = 0; i < d; i++)
    {
        sum += pow(point1[i] - point2[i], 2);
    }
    return sqrt(sum);
}

int areBiggerThanEpsilon(double *arr, int size)
{
    int i;
    for (i = 0; i < size; i++)
    {
        if (arr[i] > EPSILON)
        {
            return 1;
        }
    }
    return 0;
}

int is_integer(char *str)
{
    char *endptr;
    strtol(str, &endptr, 10);
    if (*endptr != '\0' || endptr == str)
    {
        return 0;
    }
    return 1;
}

File: kmeans.py
import sys
import math

EPSILON = 0.001

def main():

    # In case of extra/less arguments
    if (len(sys.argv) < 5 or len(sys.argv) > 6):
        print("Please Enter 4 or 5 arguments")
        exit()

    k = int(sys.argv[1])
    n = int(sys.argv[2])
    d = int(sys.argv[3])
    
    if (sys.argv[4].isdigit()):
        iter = int(sys.argv[4])
        file_name = sys.argv[5]
    else:
        # if iter is not provided
        iter = 200
        file_name = sys.argv[4]

    # Validation of the arguments
    assert n > 1, "Invalid number of points!"
    assert k > 1 and k < n, "Invalid number of clusters!"
    assert d > 1, "Invalid dimension of point!"
    assert iter > 1 and iter < 1000, "Invalid maximum iteration!"

    # Reading from file
    text_file = open(file_name, 'r')
    raw_text = text_file.read()
    text_file.close()

    # Manipulation on the data (splitting by new lines and then by comma)
    vectors = raw_text.splitlines()
    vectors_arr = [vector.split(',') for vector in vectors]

    # Step 1: Initialize first k vectors as centroids
    centroids = [vectors_arr[i] for i in range(k)]
    
    # Steps 2&5: Iterate until all delta centroids are smaller than EPSILON or max iteration is reached
    iteration = 0
    delta_centroids = [1 for _ in range(k)]
    while are_bigger_than_epsilon(delta_centroids) and iteration < iter:

        # Step 3: Assign every vector to the closest cluster 
        closest_centroid_for_vector = find_closest_centroids(vectors_arr, centroids)

        # Step 4: Update the centroids
        for centroid_index in range(k):
            sum = [0 for _ in range(d)]
            count = 0
            
            for i in range(n):
                if closest_centroid_for_vector[i] == centroid_index:
                    count += 1
                    for j in range(d):
                        sum[j] += float(vectors_arr[i][j])
                
            new_centroid = [(sum[i] / count) for i in range(d)]

            delta_centroids[centroid_index] = euclidean_distance(centroids[centroid_index], new_centroid)
            centroids[centroid_index] = new_centroid

        iteration += 1


    # Print centroids
    for centroid in centroids:
        formatted_centroid = ",".join(["{:.4f}".format(coordination) for coordination in centroid])
        print(formatted_centroid)

    return


# Function to find closest centroids for each vector
def find_closest_centroids(vectors_arr, centroids):
    closest_centroid_for_vector = []
    for vector in vectors_arr:
        distances = [euclidean_distance(vector, centroid) for centroid in centroids]
        closest_centroid_for_vector.append(distances.index(min(distances)))
    return closest_centroid_for_vector

# Function to calculate Euclidean distance
def euclidean_distance(vector_x, vector_y):
    sum = 0
    for i in range(len(vector_x)):
        sum += (float(vector_x[i]) - float(vector_y[i])) ** 2
    
    return math.sqrt(sum)

# Function to check if any value in the array is bigger than EPSILON
def are_bigger_than_epsilon(arr):
    return any(value > EPSILON for value in arr)


if __name__ == "__main__":
    main()
